githubapp/githubapp.go

```diff
@@ -0,0 +1,229 @@
+// GitHubAppConfig object.
+func New(config *GitHubAppConfig) *GitHubApp {
+	app := GitHubApp{
+		config: config,
+	}
+	if config.jwtCacheDuration != 0 {
+		app.jwtCache = cache.New[[]byte](config.jwtCacheDuration)
+	}
+	return &app
+}
+
+// AppToken creates a signed JWT to authenticate a GitHub app
+// so that it can make API calls to GitHub.
+func (g *GitHubApp) AppToken() ([]byte, error) {
+	// If token caching is enabled, look first in the cache
+	if g.jwtCache != nil {
+		// Check for a valid JWT in the cache
+		signedJwt, ok := g.jwtCache.Lookup(cacheKey)
+		if !ok {
+			// Create a JWT for reading instance information from GitHub
+			signedJwt, err := g.generateAppJWT()
+			if err != nil {
+				return nil, fmt.Errorf("error generating the JWT for GitHub app access: %w", err)
+			}
+			g.jwtCache.Set(cacheKey, signedJwt)
+		}
+		return signedJwt, nil
+	}
+	// Create a JWT for reading instance information from GitHub
+	signedJwt, err := g.generateAppJWT()
+	if err != nil {
+		return nil, fmt.Errorf("error generating the JWT for GitHub app access: %w", err)
+	}
+	return signedJwt, nil
+}
+
+// generateAppJWT builds a signed JWT that can be used to
+// communicate with GitHub as an application.
+func (g *GitHubApp) generateAppJWT() ([]byte, error) {
+	// Make the current time 30 seconds in the past to combat clock
+	// skew issues where the JWT we issue looks like it is coming
+	// from the future when it gets to GitHub
+	iat := time.Now().Add(-30 * time.Second)
+	exp := iat.Add(g.config.jwtTokenExpiration)
+	iss := g.config.AppID
+
+	token, err := jwt.NewBuilder().
+		Expiration(exp).
+		IssuedAt(iat).
+		Issuer(iss).
+		Build()
+	if err != nil {
+		return nil, fmt.Errorf("error building JWT: %w", err)
+	}
+	signed, err := jwt.Sign(token, jwt.WithKey(jwa.RS256, g.config.PrivateKey))
+	if err != nil {
+		return nil, fmt.Errorf("error signing JWT: %w", err)
+	}
+	return signed, nil
+}
+
+// AccessToken calls the GitHub API to generate a new
+// access token for this application installation with the requested
+// permissions and repositories.
+func (g *GitHubApp) AccessToken(ctx context.Context, request *TokenRequest) (string, error) {
+	appJWT, err := g.AppToken()
+	if err != nil {
+		return "", fmt.Errorf("error generating app jwt: %w", err)
+	}
+	requestURL := fmt.Sprintf(g.config.accessTokenURLPattern, g.config.InstallationID)
+	requestJSON, err := json.Marshal(request)
+	if err != nil {
+		return "", fmt.Errorf("error marshalling request data: %w", err)
+	}
+
+	requestReader := bytes.NewReader(requestJSON)
+	req, err := http.NewRequestWithContext(ctx, http.MethodPost, requestURL, requestReader)
+	if err != nil {
+		return "", fmt.Errorf("error creating http request for GitHub installation information: %w", err)
+	}
+	req.Header.Set("Accept", "application/vnd.github+json")
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", appJWT))
+
+	client := http.Client{Timeout: 10 * time.Second}
```

Comment:
Can we make this client during initialization and save it on the struct instead? It might also be good to have a `WithHTTPClient` option.

Reply:
done.